# 位元（bit）

>>1. 位元是最基本的數據單位，只有兩種值，0或1。

>>2. 在 Python 中並沒有直接處理位元的資料類型，可透過「位元運算子」來操作整數的位元（如下表）。

>>>| 符號 | 說明 |
>>>|---|---|
>>>| & | and |
>>>| \| | or |
>>>| ^ | xor |
>>>| ~ | not |
>>>| << | 左移 |
>>>| >> | 右移 |


1. and / or
>>>這裡為何 12 & 10 的結果是 8？而 12 | 10 的結果是 14 呢？

>>>要理解這點，我們需要將數字轉換為它們的二進位表示形式。在二進位中，12 可以表示為 1100，10 可以表示為 1010，而位元 AND 和位元 OR 就是對對這些位元數值進行運算。

In [7]:
# 位元 AND
result = 12 & 10
print(f'12 & 10 表示 {12:b} 與 {10:b} 的位元運算 = {result:b}') 
print(f'{result:b} 的10進位 = {result}')  # 輸出：8

print('------------------')
# 位元 OR
result = 12 | 10
print(f'12 | 10 表示 {12:b} 與 {10:b} 的位元運算 = {result:b}')
print(f'{result:b} 的10進位 = {result}')  # 輸出：14


12 & 10 表示 1100 與 1010 的位元運算 = 1000
1000 的10進位 = 8
------------------
12 | 10 表示 1100 與 1010 的位元運算 = 1110
1110 的10進位 = 14


【補充說明】

>>位元 AND（&）：兩個位元都是1，結果才為1，否則為0。

>>12: 1100
>>10: 1010
>>----------
>>&:   1000  # 這是二進位表示的 8

>>位元 OR（|）：兩個位元其中一個為1，結果就為1，兩個都為0，結果才為0。

>>12: 1100
>>10: 1010
>>----------
>>|:   1110  # 這是二進位表示的 14


# 字節（byte）

1. 一個 byte 等於8個位元(bit)，是存儲資訊的基本單位，也是最常見的電腦數據單位，因為8位元的組合可以表示256種不同的值，這對於創建更大的數字或者表示像字母和數字這樣的字符非常方便。

2. 在 Python 中，我們可以用 bytes 類型來操作字節序列。bytes 對象是由 0 到 255 之間的整數所組成的不可變序列。

3. 在字符編碼（如ASCII，UTF-8，UTF-16等）中，一個字符可能占用的位元數量可以變化。例如，在ASCII編碼中，一個字符只占用1個 byte（或8個 bit），而在UTF-16編碼中，一個字符可能占用2或4個 byte（也就是 16或32個 bit）。

4. 另外，全形字符在很多編碼系統中，例如UTF-8或UTF-16，一般都是占用2個字節的。

### 下面的範例

b 是一個 bytes 對象，它是由 5 個字節組成的。每個字節在這裡都被表示為一個介於 0 和 255 之間的整數。當我們印出 b 時，每個字節被顯示為它對應的 ASCII 字符（如果它有對應的可印字元）或其十六進位值（例如 \x0c 和 \xff）。

In [10]:
b = bytes([123, 12, 45, 78, 255])
print(b)  # 輸出：b'{\x0c-|N\xff}'
print('\n===============')
print(bytes([123]))
print(bytes([12]))
print(bytes([45]))
print(bytes([78]))
print(bytes([255]))

b'{\x0c-N\xff'

b'{'
b'\x0c'
b'-'
b'N'
b'\xff'


---END---